server/src-lib/Hasura/GraphQL/Parser/Internal/Parser.hs

```diff
@@ -0,0 +1,997 @@
+      withPath (++[Key "args"]) $ ifParser argumentsParser $ GraphQLValue <$> _fArguments
+  }
+  where
+    argumentNames = S.fromList (dName <$> ifDefinitions argumentsParser)
+
+-- | Builds a 'FieldParser' for a field that takes a subselection set, i.e. a
+-- field that returns an object.
+--
+-- See also Note [The delicate balance of GraphQL kinds] in "Hasura.GraphQL.Parser.Schema".
+subselection
+  :: forall m a b
+   . MonadParse m
+  => Name
+  -> Maybe Description
+  -> InputFieldsParser m a -- ^ parser for the input arguments
+  -> Parser 'Output m b -- ^ parser for the subselection set
+  -> FieldParser m (a, b)
+subselection name description argumentsParser bodyParser = FieldParser
+  { fDefinition = mkDefinition name description $
+      FieldInfo (ifDefinitions argumentsParser) (pType bodyParser)
+  , fParser = \Field{ _fArguments, _fSelectionSet } -> do
+      -- check for extraneous arguments here, since the InputFieldsParser just
+      -- handles parsing the fields it cares about
+      for_ (M.keys _fArguments) \argumentName -> do
+        unless (argumentName `S.member` argumentNames) $
+          parseError $ name <<> " has no argument named " <>> argumentName
+      (,) <$> withPath (++[Key "args"]) (ifParser argumentsParser $ GraphQLValue <$> _fArguments)
+          <*> pParser bodyParser _fSelectionSet
+  }
+  where
+    argumentNames = S.fromList (dName <$> ifDefinitions argumentsParser)
+
+-- | A shorthand for a 'selection' that takes no arguments.
+selection_
+  :: MonadParse m
+  => Name
+  -> Maybe Description
+  -> Parser 'Both m a -- ^ type of the result
+  -> FieldParser m ()
+selection_ name description = selection name description (pure ())
+
+-- | A shorthand for a 'subselection' that takes no arguments.
+subselection_
+  :: MonadParse m
+  => Name
+  -> Maybe Description
+  -> Parser 'Output m a -- ^ parser for the subselection set
+  -> FieldParser m a
+subselection_ name description bodyParser =
+  snd <$> subselection name description (pure ()) bodyParser
+
+
+-- -----------------------------------------------------------------------------
+-- helpers
+
+valueToJSON :: MonadParse m => GType -> InputValue Variable -> m A.Value
+valueToJSON expected = peelVariable (Just expected) >=> valueToJSON'
+  where
+    valueToJSON' = \case
+      JSONValue    j -> pure j
+      GraphQLValue g -> graphQLToJSON g
+    graphQLToJSON = \case
+      VNull               -> pure A.Null
+      VInt i              -> pure $ A.toJSON i
+      VFloat f            -> pure $ A.toJSON f
+      VString t           -> pure $ A.toJSON t
+      VBoolean b          -> pure $ A.toJSON b
+      VEnum (EnumValue n) -> pure $ A.toJSON n
+      VList values        -> A.toJSON <$> traverse graphQLToJSON values
+      VObject objects     -> A.toJSON <$> traverse graphQLToJSON objects
+      VVariable variable  -> valueToJSON' $ absurd <$> vValue variable
+
+jsonToGraphQL :: (MonadError Text m) => A.Value -> m (Value Void)
+jsonToGraphQL = \case
```

Comment:
Is this function still necessary now that we can use JSON values as `InputValue`s directly? If so, why?

Reply:
I think it is used for remote joins, to decode some parts of the remote schema such as default argument values. @codingkarthik can you confirm?

Reply:
Yes, it is used to get the value of the remote join parameter.  We get the JSON value from the DB by running the SQL with remote joins. Then, the JSON value is converted to GraphQL value. Also, we merge the input arguments (if any) which are GraphQL values with the arguments that are defined in the remote join definition.

Reply:
@codingkarthik So does that mean that some remote joins fail because a JSON object cannot be represented as GraphQL objects? If so, is that justified? I suppose an alternative model could be to pass all of these JSON joining parameters through query variables, right? (I'm not saying that that's necessarily worth the effort, or even better, I'm just trying to understand the situation.)

Reply:
> So does that mean that some remote joins fail because a JSON object cannot be represented as GraphQL objects

The only way I imagine this happening is when a table column name is not a GraphQL friendly name. 

> I suppose an alternative model could be to pass all of these JSON joining parameters through query variables,

I'm not sure how we can do this, because one of the steps is to merge the input arguments(GraphQL type) to the remote join node with the arguments retrieved from the database (while retrieved in JSON, then converted to GraphQL) and then the remote schema is queried with the merged GraphQL arguments. Suppose, we keep the joining parameters in query variables then we'll have to convert the input arguments into JSON, then merge them and then pack it into GraphQL variables and then query the remote server. 

The above approach seems doable, but on the other hand, we don't track table with columns which have non-GraphQL friendly names, so converting JSON into GraphQL values can never fail (I think..). So, I'm not sure what will be the right way to proceed further.